### ISSUE TITLE
Handling of surrogate pairs when Intl.Segmenter isn't available

### ISSUE THREAD

**Issue by johnfactotum** [arty-name/locale-index-of#19]
Sorry to trouble you, again.

Currently, when Intl.Segmenter isn't available, it uses the following to segment the string
https://github.com/arty-name/locale-index-of/blob/cb7f4005fc1c6c740bdbe1fda2e2c73e3fbda479/index.js#L58-L65

I just learned, however, that this won't handle surrogate pairs properly. For example,

```js
[...segmenter.segment('𝟘𝟙𝟚𝟛𝟜𝟝𝟞𝟟𝟠𝟡')]
```

Expected:
```js
[
  { segment: '𝟘', index: 0 },
  { segment: '𝟙', index: 2 },
  { segment: '𝟚', index: 4 },
  { segment: '𝟛', index: 6 },
  { segment: '𝟜', index: 8 },
  { segment: '𝟝', index: 10 },
  { segment: '𝟞', index: 12 },
  { segment: '𝟟', index: 14 },
  { segment: '𝟠', index: 16 },
  { segment: '𝟡', index: 18 }
]
```

Actual:
```js
[
  { segment: '\ud835', index: 0 },
  { segment: '\udfd8', index: 1 },
  { segment: '\ud835', index: 2 },
  { segment: '\udfd9', index: 3 },
  { segment: '\ud835', index: 4 },
  { segment: '\udfda', index: 5 },
  { segment: '\ud835', index: 6 },
  { segment: '\udfdb', index: 7 },
  { segment: '\ud835', index: 8 },
  { segment: '\udfdc', index: 9 },
  { segment: '\ud835', index: 10 },
  { segment: '\udfdd', index: 11 },
  { segment: '\ud835', index: 12 },
  { segment: '\udfde', index: 13 },
  { segment: '\ud835', index: 14 },
  { segment: '\udfdf', index: 15 },
  { segment: '\ud835', index: 16 },
  { segment: '\udfe0', index: 17 },
  { segment: '\ud835', index: 18 },
  { segment: '\udfe1', index: 19 }
]
```

The length and index of the string refer to code _units_, not code points. The iterator is actually the preferred method in almost all cases, as one doesn't normally want to deal with code units:
```js
const segmenter = {
  *segment(string) {
    let index = 0;
    for (const segment of string) {
      yield { segment, index };
      index += segment.length;
    }
  }
}
```

In the case of string searching, I _think_ it's actually fine to split at code units. It should still be able get a correct match.

But since it's a very easy fix I don't think there's any reason not to handle it properly, unless one wants to support IE, which doesn't support `String.prototype[@@iterator]()`...

**Comment by arty-name** [comment]
Yes, this is a totally valid point! When working on this part, I didn’t realize that, while I cannot use index for offsets in such scenario, I still absolutely need it to properly segment.

**Comment by arty-name** [comment]
I’ve published version 3.1.0 with this fix. Please check whether it works for you and close the issue if so.

Surprised that you even noticed that, I’d expect that Foliate has access to modern APIs.

**Comment by johnfactotum** [comment]
> Surprised that you even noticed that, I’d expect that Foliate has access to modern APIs.

Not in Foliate. Just something that I happened to notice while looking into the problem of string searching. I vaguely remember learning about Unicode gotchas like this in the past but never really understood it until actually running into it in practice.

For Foliate I will probably write my own implementation because it needs extra things like searching across multiple text nodes, working out the matching DOM Ranges in the document, etc.
